### ISSUE TITLE
Release plan discussion

### ISSUE THREAD

**Issue by mcuee** [mcuee/libusb-win32#26]
We will use 1.2.7.x as snapshot release version. Probably the next formal version should be labeled as 1.2.8.0.

V1.2.7.3 (11/13/2021) - SNAPSHOT RELEASE
==============================================
* driver: sign the drivers using SHA1 as well as SHA256

* driver: sign the drivers after microsoft and not before (win7 fix)

V1.2.7.2 (10/25/2021) - SNAPSHOT RELEASE
==============================================
* driver: sign the drivers using EV certificate: https://github.com/mcuee/libusb-win32/issues/24

* driver: fix possible stack corruption: https://github.com/mcuee/libusb-win32/issues/19

* lib: fix missing check for failed CloseHandle(): https://github.com/mcuee/libusb-win32/issues/12

V1.2.7.1 (09/18/2019) - SNAPSHOT RELEASE
==============================================
* Removed support for IA64
* Removed support for W2K
* Properly allocate NX pool memory on Win8+

V1.2.6.0 (01/17/2012)
==============================================
* Official release.
* Removed ISO maximum transfer size restrictions/transfer splitting.
* Fixed inf-wizard device notification issue.

**Comment by mcuee** [comment]
#25 has been merged.

**Comment by mcuee** [comment]
The following issues have been fixed. We are more or less ready for the next release.
- #19 
- #21  
- #25 
- #27

**Comment by mcuee** [comment]
Release related work to be done.

- update changelog
- including the change log in the installer
- create debug version of bin and installer
- create libusb

**Comment by mcuee** [comment]
> @dontech Please also help to check those I marked as "wontfix" before you took over from Travis. If you think some of them are worth implementing, then please go ahead to remove the "wontfix" label. Thanks.

I have removed most of the "wontfix" labels except a few like USB hub support.

**Comment by mcuee** [comment]
@dontech 
For the next release, Windows 7 support may be difficult as the cross-certification expired in Dec 2021.

```
C:\Program Files (x86)\Windows Kits\10\bin\10.0.19041.0\x64> .\signtool.exe verify /v /kp C:\temp\libusb-win32-bin-1.2.7.3\bin\amd64\libusb0.sys

Verifying: C:\temp\libusb-win32-bin-1.2.7.3\bin\amd64\libusb0.sys

Signature Index: 0 (Primary Signature)
Hash of file (sha256): 2148A71E5EF84643B11A1A528E421B3860D2D07831FB7BCECA9D0E5DE7B4AC43

Signing Certificate Chain:
Issued to: Microsoft Root Certificate Authority 2010
Issued by: Microsoft Root Certificate Authority 2010
Expires: Sun Jun 24 06:04:01 2035
SHA1 hash: 3B1EFD3A66EA28B16697394703A72CA340A05BD5

    Issued to: Microsoft Windows Third Party Component CA 2014
    Issued by: Microsoft Root Certificate Authority 2010
    Expires:   Tue Oct 16 04:41:27 2029
    SHA1 hash: 1906DCF62629B563252C826FDD874EFCEB6856C6

        Issued to: Microsoft Windows Hardware Compatibility Publisher
        Issued by: Microsoft Windows Third Party Component CA 2014
        Expires:   Fri Dec 03 06:25:28 2021
        SHA1 hash: 984E03B613E8C2AE9C692F0DB2C031BF3EE3A0FA
The signature is timestamped: Sat Nov 13 19:54:14 2021
Timestamp Verified by:
Issued to: Microsoft Root Certificate Authority 2010
Issued by: Microsoft Root Certificate Authority 2010
Expires: Sun Jun 24 06:04:01 2035
SHA1 hash: 3B1EFD3A66EA28B16697394703A72CA340A05BD5

    Issued to: Microsoft Time-Stamp PCA 2010
    Issued by: Microsoft Root Certificate Authority 2010
    Expires:   Wed Jul 02 05:46:55 2025
    SHA1 hash: 2AA752FE64C49ABE82913C463529CF10FF2F04EE

        Issued to: Microsoft Time-Stamp Service
        Issued by: Microsoft Time-Stamp PCA 2010
        Expires:   Thu Jan 13 01:28:25 2022
        SHA1 hash: A04C15E8F4C630938C2935B1375329AF93CC90D4
Cross Certificate Chain:
Issued to: Microsoft Root Certificate Authority 2010
Issued by: Microsoft Root Certificate Authority 2010
Expires: Sun Jun 24 06:04:01 2035
SHA1 hash: 3B1EFD3A66EA28B16697394703A72CA340A05BD5

    Issued to: Microsoft Windows Third Party Component CA 2014
    Issued by: Microsoft Root Certificate Authority 2010
    Expires:   Tue Oct 16 04:41:27 2029
    SHA1 hash: 1906DCF62629B563252C826FDD874EFCEB6856C6

        Issued to: Microsoft Windows Hardware Compatibility Publisher
        Issued by: Microsoft Windows Third Party Component CA 2014
        Expires:   Fri Dec 03 06:25:28 2021
        SHA1 hash: 984E03B613E8C2AE9C692F0DB2C031BF3EE3A0FA
Successfully verified: C:\temp\libusb-win32-bin-1.2.7.3\bin\amd64\libusb0.sys

Number of files successfully Verified: 1
Number of warnings: 0
Number of errors: 0

```

**Comment by mcuee** [comment]
Further discussions here.
https://community.osr.com/discussion/293709/signing-driver-for-win7

**Comment by dontech** [comment]
@mcuee yes there clearly is a problem there. I have not been able to solve this either.

Either i am missing something or Microsoft made a huge backwards compability support blunder, which is problematic when you consider how many people still use that....

AFAIK drivers for Windows 7 can _only_ be signed using their cross-certificate, and since they have stopped issueing those, then i fail to see how that would work.

They need to issue a new cross-certificate or patch up Windows-7 to ignore that check.... i guess the odds of any of those two are smaller than my odds of becoming the president of Mars.

So i think, unless something happens, that we have to drop support for Windows 7.

**Comment by dontech** [comment]
I would suggest that we announce that this is the last version supporting Windows 7, and then remove it in the next release. We can state in the release notes that it must be signed manually (local signing), or installed using the "disable driver signing" options.

**Comment by mcuee** [comment]
@dontech 
I agree with you to annonuce libusb-win32 1.2.7.3 to be the last version to support Windows 7.

I have updated the github release page. 
https://github.com/mcuee/libusb-win32/releases/tag/snapshot_1.2.7.3

Or probably we should re-tag it as official 1.2.7.3 release and not call it  a snapshot.

**Comment by mcuee** [comment]
@dontech 

I have just upload 1.2.7.3 again to Sourceforge as a formal release.
https://sourceforge.net/projects/libusb-win32/files/libusb-win32-releases/1.2.7.3/

**Comment by mcuee** [comment]
1.2.7.4 snapshot release:
https://sourceforge.net/projects/libusb-win32/files/libusb-win32-snapshots/20230920/
https://github.com/mcuee/libusb-win32/releases/tag/snapshot_1.2.7.4

**Comment by mcuee** [comment]
@dontech 

Interesting that I saw SHA1 signature on the 1.2.7.4 release along with SHA256 signature. Are you able to sort the issue for Windows 7? Does 1.2.7.4 work under Windows 7? Thanks.

**Comment by dontech** [comment]
What do we need to complete this.

@mcuee what do you think?

**Comment by mcuee** [comment]
> What do we need to complete this.
> 
> @mcuee what do you think?

I think we are almost there.

Just two minor things for the installer.
1) It would be nice to sign the installer. I think it will help to make Windows a bit happy.

![Screenshot 2023-10-05 120138](https://github.com/mcuee/libusb-win32/assets/1294740/dbe4fd41-06b1-4c76-986f-cbb81902ddf9)

2) Add the changelog to the installer.

![Screenshot 2023-10-05 120240](https://github.com/mcuee/libusb-win32/assets/1294740/e8759562-fc0b-4541-8dcb-1701baf8fe04)

**Comment by mcuee** [comment]
After the formal release, I will create a pull request to ask Pete to include it in the next official release of libwdi/Zadig.

Currently libwdi Zadig-2.8.exe has libusb-win32 1.2.7.3 release.
* https://github.com/pbatard/libwdi/pull/282

**Comment by mcuee** [comment]
@dontech 

I can not test Windows on ARM64 as I do not have the necessary hardware. I am not so sure if you can test Windows on ARM64 or not.

There is an issue with Zadig as of now. It does not work under Windows 11 on ARM64. Power users can of course disable driver signature enforcement, but this is not a solution for normal users.
* https://github.com/pbatard/libwdi/issues/289

**Comment by tormodvolden** [comment]
Why does the changelog say "Much smaller binaries"? At least in the case of amd64 all binaries are larger in 1.3.0.0 than in 1.2.7.4...

**Comment by dontech** [comment]
> Why does the changelog say "Much smaller binaries"? At least in the case of amd64 all binaries are larger in 1.3.0.0 than in 1.2.7.4...

@tormodvolden That's a really good question. At some point I noticed that the binaries were smaller when switching to EWDK. I just never checked it again.

Maybe I configured the project differently at some point. Not sure. I think its OK now, but I agree that the release note is misleading. Maybe some of the project parameters can be tweaked.

**Comment by dontech** [comment]
> 1. It would be nice to sign the installer. I think it will help to make Windows a bit happy.

Done. The latest update to the script signs all EXE and DLL files with SHA1 and SHA256 signature.
The SYS and DLL files then additionally get signed by microsoft.

**Comment by dontech** [comment]
![image](https://github.com/mcuee/libusb-win32/assets/5517911/55392bd3-8888-4a67-8481-5193260d8fe3)

**Comment by dontech** [comment]
CHANGELOG.txt now a part of the archive, and should be updated with every release.

**Comment by dontech** [comment]
@mcuee We are ready for release. Please confirm and I will spin and sign test release

**Comment by tormodvolden** [comment]
The changelog still has the dubious "Much smaller binaries due to newer compiler" ...

**Comment by mcuee** [comment]
> @mcuee We are ready for release. Please confirm and I will spin and sign test release

Thanks. Please go ahead with the new release. But maybe you can remove "Much smaller binaries due to newer compiler" as mentioned by Tormod.

**Comment by dontech** [comment]
Going back and changing a changelog is normally a no-go, but in this case we can make an exception as it is misleading.

**Comment by mcuee** [comment]
@dontech 

Now it should be good to go. Thanks.

**Comment by tormodvolden** [comment]
A changelog is documentation and should be corrected and improved if needed. If you have a strict QM system you may need to document that the documentation has been changed, but I never heard about such a no-go rule :)

**Comment by dontech** [comment]
> A changelog is documentation and should be corrected and improved if needed. If you have a strict QM system you may need to document that the documentation has been changed, but I never heard about such a no-go rule :)

I tried to say the same thing, but your wording is better. I agree. :-).

**Comment by dontech** [comment]
https://github.com/mcuee/libusb-win32/releases/tag/snapshot_1.3.0.1

Ready to test!

**Comment by dontech** [comment]
One interesting thing could be to test the ARM64 images somehow.

**Comment by mcuee** [comment]
> One interesting thing could be to test the ARM64 images somehow.

@sonatique tested 1.3.0.0 snapshot release under ARM64 and it seems to work fine. I think 1.3.0.1 shoudl be similar.
https://github.com/libusb/libusb/issues/1255#issuecomment-1759140351

**Comment by mcuee** [comment]
Let's wait for some time and we may want to promote 1.3.0.1 snapshot release to be a formal release, or just up-rev and create a 1.3.0.2 as formal release.

**Comment by mcuee** [comment]
@dontech 

Strange that I still see the installer as unsigned and I still see the pop-up warnning.

You can see no digital signature tab for the file.
![Screenshot 2023-10-28 073653](https://github.com/mcuee/libusb-win32/assets/1294740/6b4deabc-c6b6-4e39-a739-70d3d8356f47)

The pop-up warning message is still there.
![Screenshot 2023-10-28 073420](https://github.com/mcuee/libusb-win32/assets/1294740/ca11aca6-09e6-4920-b9a5-92b1bb3f861d)

**Comment by mcuee** [comment]
@dontech 

Example of a signed application (libwdi Zadig). I think you can sign with your certificate, no need to go through Microsoft since this is not a driver.
![Screenshot 2023-10-28 074008](https://github.com/mcuee/libusb-win32/assets/1294740/e7e6ac83-5a4b-4168-97ca-eb34c13362d6)

**Comment by mcuee** [comment]
Another thing for 1.3.0.2 is probably to document the limitation for ARM64.

As of now, the work-around is to submit the generated Zadig driver package to Microsoft portal (attestation signing) or to Disable Driver Signature Enforcement.
* https://github.com/pbatard/libwdi/issues/289

**Comment by mcuee** [comment]
Yet another thing for 1.3.0.2 is to fix the issue with Zadig (but it may be a Zadig problem and not libusb-win32 problem).
* https://github.com/pbatard/libwdi/issues/301

**Comment by mcuee** [comment]
Not sure if we can adopt the following.
https://learn.microsoft.com/en-us/windows-hardware/drivers/devtest/static-tools-and-codeql#driver-verification-log-dvl-consumption-of-sarif-output)

**Comment by mcuee** [comment]
```
V1.3.0.2 (02/15/2024) - SNAPSHOT RELEASE

- Fix Linux build: https://github.com/mcuee/libusb-win32/issues/42
- Fix missing data in large transfers: https://github.com/mcuee/libusb-win32/issues/45
- Fix BSOD for large transfers: https://github.com/mcuee/libusb-win32/issues/51
- Replace pipeinfo transfer size with MS rules: https://github.com/mcuee/libusb-win32/issues/52
- Fix async API data order: https://github.com/mcuee/libusb-win32/issues/54
- Fix missing PDB files for release: https://github.com/mcuee/libusb-win32/issues/55
- Fix MINGW build: https://github.com/mcuee/libusb-win32/issues/58

V1.3.0.1 (10/20/2023) - SNAPSHOT RELEASE
Automate the driver signing process more
Sign all executables
Fix changelog in build system
```

**Comment by mcuee** [comment]
@dontech 

BTW, version 1.3.0.2 installer is missing the changelog for 1.3.0.2. You may want to add that in the next version.

**Comment by dontech** [comment]
> BTW, version 1.3.0.2 installer is missing the changelog for 1.3.0.2. You may want to add that in the next version.

Yeah i committed it _after_ release, and obviously is should do it _before_ release... doh! :-)

**Comment by dontech** [comment]
If there are no new problems, I think we are ready to release 1.4.0.0 as the next _official_ release

**Comment by mcuee** [comment]
> If there are no new problems, I think we are ready to release 1.4.0.0 as the next _official_ release

Yes, it will be good to have a real official release.

**Comment by mcuee** [comment]
@dontech 

There are four old enhancement requests. If you think they are not worth the efforts, we can close them as not planned.
* #1 
* #2 
* #10 
* #11

**Comment by sonatique** [comment]
Congratulation and big thanks to @dontech and @mcuee for the newer 1.4.0.0 release!

**Comment by mcuee** [comment]
@sonatique 

Thanks a lot for your contributions as well.

**Comment by mcuee** [comment]
@dontech 

Just wondering if you can add the "libusb-win32-devel-filter-1.4.0.0.exe" installer release as well. Thanks.

**Comment by mcuee** [comment]
@dontech 

Ping for the installer release. Thanks.

**Comment by devops-dmdx** [comment]
Hi,

Thanks for the great work!

While exploring the content of release 1.4.0.0, i noticed to new files in the `bin` folder:
- `libusb0.inf`
- `libusb0.cat` 

The INF-file mentions `DeviceID   = "VID_0B05&PID_190E"`, so my guess is that this INF + catalog file serves as a template for installing the driver for our own devices, am i right? 

Is the purpose of these two files documented somewhere but I couldn't find it?

Best,
Hugo

**Comment by mcuee** [comment]
You can ignore the two files. It is part of the attestation signing package to get the signed libusb0.sys from Microsoft portal.

You should use Zadig 2.9 release to install libusb-win32 device driver for your device.
https://github.com/pbatard/libwdi/releases/tag/v1.5.1

The inf file is useful for people who want to sign their own driver packages (if they have the necessary EV digital certiticate).

**Comment by devops-dmdx** [comment]
Thanks for the prompt reply. This is clear now.

Btw, is the -devel-filter.exe scheduled for release somewhen?

**Comment by mcuee** [comment]
> Thanks for the prompt reply. This is clear now.
> 
> Btw, is the -devel-filter.exe scheduled for release somewhen?

We have to wait for the answer from @dontech.

**Comment by mcuee** [comment]
> [@dontech](https://github.com/dontech) Ping about ` -devel-filter.exe` for 1.4.0. Thanks.

Done with latest 1.4.0.2 release.
https://github.com/mcuee/libusb-win32/releases/tag/release_1.4.0.2

**Comment by mcuee** [comment]
> [@dontech](https://github.com/dontech)
> 
> There are four old enhancement requests. If you think they are not worth the efforts, we can close them as not planned.
> 
> * [Feature request: usb_get_configuration #1](https://github.com/mcuee/libusb-win32/issues/1)
> * [Feature request:  isochronous buffer packets offset/length #2](https://github.com/mcuee/libusb-win32/issues/2)
> * [Enhancement with patch: Exposing usb overlapped event #10](https://github.com/mcuee/libusb-win32/issues/10)
> * [Enhancement: Making a distinction between ERROR_SEM_TIMEOUT and ERROR_OPERATION_ABORTED #11](https://github.com/mcuee/libusb-win32/issues/11)

Closed them as not planned.
